File: explainer/src/lib/engine/entity.ts
import {
  type EntityId,
  type NameId,
  type Tick,
  type EntityState,
  LifecycleState,
  isActiveState,
} from './types';

/**
 * A timestamped state change for an entity.
 * The core building block - we store events, not snapshots.
 */
export interface Event<S extends EntityState> {
  timestamp: Tick;
  lifecycle: LifecycleState;
  state: S;
}

/**
 * Create a new event with the Born lifecycle state
 */
export function bornEvent<S extends EntityState>(timestamp: Tick, state: S): Event<S> {
  return { timestamp, lifecycle: LifecycleState.Born, state };
}

/**
 * Create a new event with the Prebirth lifecycle state
 */
export function prebirthEvent<S extends EntityState>(timestamp: Tick, state: S): Event<S> {
  return { timestamp, lifecycle: LifecycleState.Prebirth, state };
}

/**
 * Create a new event with the Dead lifecycle state
 */
export function deadEvent<S extends EntityState>(timestamp: Tick, state: S): Event<S> {
  return { timestamp, lifecycle: LifecycleState.Dead, state };
}

/**
 * Create a new event with the Chronoporting lifecycle state
 */
export function chronoportingEvent<S extends EntityState>(timestamp: Tick, state: S): Event<S> {
  return { timestamp, lifecycle: LifecycleState.Chronoporting, state };
}

/**
 * An entity with its complete event history.
 *
 * Key design: State is stored as a list of timestamped events, not full snapshots.
 * This enables O(log n) historical queries via binary search.
 */
export class Entity<S extends EntityState> {
  /** Unique identifier */
  readonly id: EntityId;

  /** Links temporal duplicates (same NameId = same "person" across time) */
  readonly nameId: NameId;

  /** When this entity was first created */
  readonly createdAt: Tick;

  /** Event history, always sorted by timestamp */
  private events: Event<S>[];

  constructor(id: EntityId, nameId: NameId, createdAt: Tick, initialEvent: Event<S>) {
    this.id = id;
    this.nameId = nameId;
    this.createdAt = createdAt;
    this.events = [initialEvent];
  }

  /**
   * Create a new entity that starts as Born at the given tick.
   */
  static spawn<S extends EntityState>(
    id: EntityId,
    nameId: NameId,
    tick: Tick,
    state: S
  ): Entity<S> {
    return new Entity(id, nameId, tick, bornEvent(tick, state));
  }

  /**
   * Create a new entity that starts as Prebirth (scheduled to appear).
   */
  static spawnPrebirth<S extends EntityState>(
    id: EntityId,
    nameId: NameId,
    tick: Tick,
    state: S
  ): Entity<S> {
    return new Entity(id, nameId, tick, prebirthEvent(tick, state));
  }

  /**
   * Get the event at or before the given tick.
   * Uses binary search for O(log n) complexity.
   */
  getEventAt(tick: Tick): Event<S> | null {
    if (this.events.length === 0) return null;
    if (tick < this.events[0].timestamp) return null;

    // Binary search for the rightmost event <= tick
    let left = 0;
    let right = this.events.length - 1;

    while (left < right) {
      const mid = Math.ceil((left + right + 1) / 2);
      if (this.events[mid].timestamp <= tick) {
        left = mid;
      } else {
        right = mid - 1;
      }
    }

    return this.events[left];
  }

  /**
   * Get the state at a specific tick.
   */
  getStateAt(tick: Tick): S | null {
    const event = this.getEventAt(tick);
    return event?.state ?? null;
  }

  /**
   * Get the lifecycle state at a specific tick.
   */
  getLifecycleAt(tick: Tick): LifecycleState {
    const event = this.getEventAt(tick);
    return event?.lifecycle ?? LifecycleState.Unborn;
  }

  /**
   * Check if the entity is active (Born or Prebirth) at the given tick.
   */
  isActiveAt(tick: Tick): boolean {
    return isActiveState(this.getLifecycleAt(tick));
  }

  /**
   * Get the most recent event.
   */
  latestEvent(): Event<S> {
    return this.events[this.events.length - 1];
  }

  /**
   * Get all events (for visualization).
   */
  allEvents(): readonly Event<S>[] {
    return this.events;
  }

  /**
   * Add an event, maintaining sorted order.
   */
  addEvent(event: Event<S>): void {
    // Find insertion point
    let insertIndex = this.events.length;
    for (let i = this.events.length - 1; i >= 0; i--) {
      if (this.events[i].timestamp <= event.timestamp) {
        insertIndex = i + 1;
        break;
      }
      if (i === 0) {
        insertIndex = 0;
      }
    }
    this.events.splice(insertIndex, 0, event);
  }

  /**
   * Set the entity's state at the given tick (creates a Born event).
   */
  setState(tick: Tick, state: S): void {
    this.addEvent(bornEvent(tick, state));
  }

  /**
   * Mark the entity as dead at the given tick.
   */
  destroy(tick: Tick): void {
    const currentState = this.getStateAt(tick);
    if (currentState) {
      this.addEvent(deadEvent(tick, currentState.clone() as S));
    }
  }

  /**
   * Mark the entity as chronoporting at the given tick.
   */
  markChronoporting(tick: Tick): void {
    const currentState = this.getStateAt(tick);
    if (currentState) {
      this.addEvent(chronoportingEvent(tick, currentState.clone() as S));
    }
  }

  /**
   * Check if the entity needs activation at the given tick.
   * Returns true if the entity is in Prebirth state at this tick.
   */
  needsActivationAt(tick: Tick): boolean {
    const event = this.getEventAt(tick);
    return event !== null && event.lifecycle === LifecycleState.Prebirth;
  }

  /**
   * Activate a Prebirth entity at the given tick (transition to Born).
   */
  activateAt(tick: Tick): void {
    const currentState = this.getStateAt(tick);
    if (currentState && this.needsActivationAt(tick)) {
      this.addEvent(bornEvent(tick, currentState.clone() as S));
    }
  }

  /**
   * Serialize the entity to a plain object for storage.
   */
  serialize(): {
    id: EntityId;
    nameId: NameId;
    createdAt: Tick;
    events: { timestamp: Tick; lifecycle: LifecycleState; state: unknown }[];
  } {
    return {
      id: this.id,
      nameId: this.nameId,
      createdAt: this.createdAt,
      events: this.events.map(e => ({
        timestamp: e.timestamp,
        lifecycle: e.lifecycle,
        state: e.state,
      })),
    };
  }
}

File: src/entity.rs
//! Entity and event types

use crate::{LifecycleState, Tick};

/// Unique identifier for an entity
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct EntityId(pub u32);

/// Identifier linking temporal duplicates ("same name" in original engine)
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct NameId(pub u32);

/// Trait for entity state data
///
/// Implement this for your game's unit/entity state.
pub trait EntityState: Clone + Default {}

/// A single event (state snapshot) in an entity's timeline
#[derive(Debug, Clone)]
pub struct Event<S: EntityState> {
    /// When this event occurs
    pub timestamp: Tick,
    /// Entity lifecycle state
    pub lifecycle: LifecycleState,
    /// The actual state data
    pub state: S,
}

impl<S: EntityState> Event<S> {
    /// Create a new event
    pub fn new(timestamp: Tick, lifecycle: LifecycleState, state: S) -> Self {
        Self {
            timestamp,
            lifecycle,
            state,
        }
    }

    /// Create a "born" event (normal spawn)
    pub fn born(timestamp: Tick, state: S) -> Self {
        Self::new(timestamp, LifecycleState::Born, state)
    }

    /// Create a "prebirth" event (scheduled to appear)
    pub fn prebirth(timestamp: Tick, state: S) -> Self {
        Self::new(timestamp, LifecycleState::Prebirth, state)
    }

    /// Create a "dead" event
    pub fn dead(timestamp: Tick, state: S) -> Self {
        Self::new(timestamp, LifecycleState::Dead, state)
    }

    /// Create a "chronoporting" event
    pub fn chronoporting(timestamp: Tick, state: S) -> Self {
        Self::new(timestamp, LifecycleState::Chronoporting, state)
    }

    /// Check if entity is active at this event
    pub fn is_active(&self) -> bool {
        matches!(
            self.lifecycle,
            LifecycleState::Born | LifecycleState::Prebirth
        )
    }
}

/// An entity with its event history
#[derive(Debug)]
pub struct Entity<S: EntityState> {
    /// Unique identifier
    pub id: EntityId,
    /// Name ID for same-name linking (temporal duplicates share this)
    pub name_id: NameId,
    /// Creation time
    pub created_at: Tick,
    /// Event history (sorted by timestamp)
    events: Vec<Event<S>>,
}

impl<S: EntityState> Entity<S> {
    /// Create a new entity
    pub fn new(id: EntityId, name_id: NameId, created_at: Tick, initial_state: S) -> Self {
        let initial_event = Event::born(created_at, initial_state);
        Self {
            id,
            name_id,
            created_at,
            events: vec![initial_event],
        }
    }

    /// Create an entity scheduled to appear at a future time
    pub fn prebirth(id: EntityId, name_id: NameId, appear_at: Tick, state: S) -> Self {
        let event = Event::prebirth(appear_at, state);
        Self {
            id,
            name_id,
            created_at: appear_at,
            events: vec![event],
        }
    }

    /// Get the most recent event at or before the given time
    pub fn get_event_at(&self, tick: Tick) -> Option<&Event<S>> {
        // Binary search for efficiency with large histories
        let idx = self
            .events
            .partition_point(|e| e.timestamp <= tick)
            .saturating_sub(1);

        self.events.get(idx).filter(|e| e.timestamp <= tick)
    }

    /// Get the most recent event
    pub fn latest_event(&self) -> Option<&Event<S>> {
        self.events.last()
    }

    /// Add a new event to the history
    pub fn add_event(&mut self, event: Event<S>) {
        // Insert in sorted order by timestamp
        let pos = self.events.partition_point(|e| e.timestamp < event.timestamp);
        self.events.insert(pos, event);
    }

    /// Add a state change at the given time
    pub fn set_state(&mut self, tick: Tick, state: S) {
        self.add_event(Event::born(tick, state));
    }

    /// Mark entity as dead at the given time
    pub fn destroy(&mut self, tick: Tick) {
        if let Some(last) = self.events.last() {
            self.add_event(Event::dead(tick, last.state.clone()));
        }
    }

    /// Mark entity as chronoporting
    pub fn mark_chronoporting(&mut self, tick: Tick) {
        if let Some(last) = self.events.last() {
            self.add_event(Event::chronoporting(tick, last.state.clone()));
        }
    }

    /// Check if entity needs activation at the given tick
    ///
    /// Returns true if:
    /// - Entity has a Prebirth event at or before tick
    /// - No Born event exists after that Prebirth event
    pub fn needs_activation_at(&self, tick: Tick) -> bool {
        if let Some(event) = self.get_event_at(tick) {
            if event.lifecycle == LifecycleState::Prebirth {
                // Check if there's a subsequent Born event
                let has_born_after = self.events.iter().any(|e| {
                    e.timestamp > event.timestamp && e.lifecycle == LifecycleState::Born
                });
                return !has_born_after;
            }
        }
        false
    }

    /// Activate the entity (transition Prebirth → Born) at the given tick
    ///
    /// This is called when a timewave reaches/passes an entity's prebirth time.
    /// In the original engine, this transitions units from scheduled to active.
    pub fn activate_at(&mut self, tick: Tick) {
        if let Some(event) = self.get_event_at(tick) {
            if event.lifecycle == LifecycleState::Prebirth {
                // Clone state and create Born event
                self.add_event(Event::born(tick, event.state.clone()));
            }
        }
    }

    /// Get the number of events in history
    pub fn event_count(&self) -> usize {
        self.events.len()
    }

    /// Iterate over all events
    pub fn events(&self) -> impl Iterator<Item = &Event<S>> {
        self.events.iter()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Clone, Default, Debug, PartialEq)]
    struct TestState {
        value: i32,
    }

    impl EntityState for TestState {}

    #[test]
    fn test_event_at_time() {
        let mut entity = Entity::new(
            EntityId(0),
            NameId(0),
            0,
            TestState { value: 0 },
        );

        entity.set_state(5, TestState { value: 50 });
        entity.set_state(10, TestState { value: 100 });

        // At t=0, should see value=0
        assert_eq!(entity.get_event_at(0).unwrap().state.value, 0);

        // At t=3, should still see value=0 (before t=5 change)
        assert_eq!(entity.get_event_at(3).unwrap().state.value, 0);

        // At t=7, should see value=50
        assert_eq!(entity.get_event_at(7).unwrap().state.value, 50);

        // At t=15, should see value=100
        assert_eq!(entity.get_event_at(15).unwrap().state.value, 100);
    }
}
